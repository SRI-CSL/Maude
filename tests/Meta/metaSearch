#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Mixfix
export MAUDE_LIB

../../src/Mixfix/maude \
  < $srcdir/metaSearch.maude -no-banner \
  > metaSearch.out 2>&1

diff $srcdir/metaSearch.expected metaSearch.out > /dev/null 2>&1
