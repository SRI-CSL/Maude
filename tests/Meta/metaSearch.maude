set show timing off .
set show advisories off .

mod SEARCH-TEST is
  sort Nat .
  op _+_ : Nat Nat -> Nat [assoc comm prec 5] .
  op _*_ : Nat Nat -> Nat [assoc comm prec 3] .
  ops a b c d e 0 1 : -> Nat .
  vars W X Y Z : Nat .
  eq 0 * X = 0 .
  eq 1 * X = X .
  eq 0 + X = X .
  rl X * (Y + Z) => X * Y + X * Z .
  rl X * Y + X * Z => X * (Y + Z) .
endm

select META-LEVEL .

red metaSearch(['SEARCH-TEST],
    '_*_['a.Nat, '_+_['b.Nat, 'c.Nat]],
    'X:Nat,
    '_*_['Y:Nat, 'Z:Nat] := 'X:Nat,
    '*,
    100,
    0) .

red metaSearch(['SEARCH-TEST],
    '_*_['a.Nat, '_+_['b.Nat, 'c.Nat]],
    'X:Nat,
    '_*_['Y:Nat, 'Z:Nat] := 'X:Nat,
    '*,
    100,
    1) .

red metaSearch(['SEARCH-TEST],
    '_*_['a.Nat, '_+_['b.Nat, 'c.Nat]],
    'X:Nat,
    '_*_['Y:Nat, 'Z:Nat] := 'X:Nat,
    '*,
    100,
    2) .

red metaSearch(['SEARCH-TEST],
    '_*_['a.Nat, '_+_['b.Nat, 'c.Nat]],
    'X:Nat,
    '_*_['Y:Nat, 'Z:Nat] := 'X:Nat /\ 'Y:Nat = 'a.Nat,
    '*,
    100,
    0) .

red metaSearch(['SEARCH-TEST],
    '_*_['a.Nat, '_+_['b.Nat, 'c.Nat]],
    'X:Nat,
    '_*_['Y:Nat, 'Z:Nat] := 'X:Nat /\ 'Y:Nat = 'a.Nat,
    '*,
    100,
    1) .

red metaSearch(['SEARCH-TEST],
    '_*_['a.Nat, '_+_['b.Nat, 'c.Nat]],
    'X:Nat,
    nil,
    '*,
    100,
    0) .

red metaSearch(['SEARCH-TEST],
    '_*_['a.Nat, '_+_['b.Nat, 'c.Nat]],
    'X:Nat,
    nil,
    '*,
    100,
    1) .

red metaSearch(['SEARCH-TEST],
    '_*_['a.Nat, '_+_['b.Nat, 'c.Nat]],
    'X:Nat,
    nil,
    '*,
    100,
    2) .

red metaSearch(['SEARCH-TEST],
    '_*_['a.Nat, '_+_['b.Nat, 'c.Nat]],
    'X:Nat,
    nil,
    '!,
    100,
    0) .

red metaSearch(['SEARCH-TEST],
    '_*_['a.Nat, '_+_['b.Nat, 'c.Nat]],
    'X:Nat,
    nil,
    '+,
    100,
    0) .

red metaSearch(['SEARCH-TEST],
    '_*_['a.Nat, '_+_['b.Nat, 'c.Nat]],
    'X:Nat,
    nil,
    '+,
    100,
    1) .

red metaSearch(['SEARCH-TEST],
   '_+_['_*_['a.Nat,'b.Nat],'_*_['a.Nat,'c.Nat]],
   '_*_['X:Nat, 'Y:Nat],
   nil,
   '*,
   100,
   0) .

red metaSearch(['SEARCH-TEST],
   '_+_['_*_['a.Nat,'b.Nat],'_*_['a.Nat,'c.Nat]],
   '_*_['X:Nat, 'Y:Nat],
   nil,
   '*,
   100,
   1) .

red metaSearch(['SEARCH-TEST],
   '_+_['_*_['a.Nat,'b.Nat],'_*_['a.Nat,'c.Nat]],
   '_*_['X:Nat, 'Y:Nat],
   nil,
   '*,
   0,
   0) .

red metaSearch(['SEARCH-TEST],
   '_+_['_*_['a.Nat,'b.Nat],'_*_['a.Nat,'c.Nat], '_*_['a.Nat,'d.Nat]],
   '_*_['X:Nat, 'Y:Nat],
   nil,
   '*,
   100,
   0) .

red metaSearch(['SEARCH-TEST],
   '_+_['_*_['a.Nat,'b.Nat],'_*_['a.Nat,'c.Nat], '_*_['a.Nat,'d.Nat]],
   '_*_['X:Nat, 'Y:Nat],
   nil,
   '*,
   100,
   1) .

red metaSearch(['SEARCH-TEST],
   '_+_['_*_['a.Nat,'b.Nat],'_*_['a.Nat,'c.Nat], '_*_['a.Nat,'d.Nat]],
   '_*_['X:Nat, 'Y:Nat],
   nil,
   '*,
   1,
   0) .

red metaSearch(['SEARCH-TEST],
   '_+_['_*_['a.Nat,'b.Nat],'_*_['a.Nat,'c.Nat], '_*_['a.Nat,'d.Nat]],
   '_*_['X:Nat, 'Y:Nat],
   nil,
   '*,
   1,
   1) .
