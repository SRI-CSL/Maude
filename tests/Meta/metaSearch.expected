Maude> ==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_*_['a.Nat,'_+_['b.Nat,
    'c.Nat]], 'X:Nat, '_*_['Y:Nat,'Z:Nat] := 'X:Nat, '*, 100, 0) .
rewrites: 2
result ResultTriple: {'_*_['a.Nat,'_+_['b.Nat,'c.Nat]],'Nat,
	'X:Nat <- '_*_['a.Nat,'_+_['b.Nat,'c.Nat]] ; 
	'Y:Nat <- '_+_['b.Nat,'c.Nat] ; 
	'Z:Nat <- 'a.Nat}
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_*_['a.Nat,'_+_['b.Nat,
    'c.Nat]], 'X:Nat, '_*_['Y:Nat,'Z:Nat] := 'X:Nat, '*, 100, 1) .
rewrites: 2
result ResultTriple: {'_*_['a.Nat,'_+_['b.Nat,'c.Nat]],'Nat,
	'X:Nat <- '_*_['a.Nat,'_+_['b.Nat,'c.Nat]] ; 
	'Y:Nat <- 'a.Nat ; 
	'Z:Nat <- '_+_['b.Nat,'c.Nat]}
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_*_['a.Nat,'_+_['b.Nat,
    'c.Nat]], 'X:Nat, '_*_['Y:Nat,'Z:Nat] := 'X:Nat, '*, 100, 2) .
rewrites: 6
result ResultTriple?: (failure).ResultTriple?
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_*_['a.Nat,'_+_['b.Nat,
    'c.Nat]], 'X:Nat, '_*_['Y:Nat,'Z:Nat] := 'X:Nat /\ 'Y:Nat = 'a.Nat, '*,
    100, 0) .
rewrites: 2
result ResultTriple: {'_*_['a.Nat,'_+_['b.Nat,'c.Nat]],'Nat,
	'X:Nat <- '_*_['a.Nat,'_+_['b.Nat,'c.Nat]] ; 
	'Y:Nat <- 'a.Nat ; 
	'Z:Nat <- '_+_['b.Nat,'c.Nat]}
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_*_['a.Nat,'_+_['b.Nat,
    'c.Nat]], 'X:Nat, '_*_['Y:Nat,'Z:Nat] := 'X:Nat /\ 'Y:Nat = 'a.Nat, '*,
    100, 1) .
rewrites: 6
result ResultTriple?: (failure).ResultTriple?
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_*_['a.Nat,'_+_['b.Nat,
    'c.Nat]], 'X:Nat, nil, '*, 100, 0) .
rewrites: 2
result ResultTriple: {'_*_['a.Nat,'_+_['b.Nat,'c.Nat]],'Nat,
	'X:Nat <- '_*_['a.Nat,'_+_['b.Nat,'c.Nat]]}
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_*_['a.Nat,'_+_['b.Nat,
    'c.Nat]], 'X:Nat, nil, '*, 100, 1) .
rewrites: 3
result ResultTriple: {'_+_['_*_['a.Nat,'b.Nat],'_*_['a.Nat,'c.Nat]],'Nat,
	'X:Nat <- '_+_['_*_['a.Nat,'b.Nat],'_*_['a.Nat,'c.Nat]]}
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_*_['a.Nat,'_+_['b.Nat,
    'c.Nat]], 'X:Nat, nil, '*, 100, 2) .
rewrites: 5
result ResultTriple?: (failure).ResultTriple?
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_*_['a.Nat,'_+_['b.Nat,
    'c.Nat]], 'X:Nat, nil, '!, 100, 0) .
rewrites: 6
result ResultTriple?: (failure).ResultTriple?
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_*_['a.Nat,'_+_['b.Nat,
    'c.Nat]], 'X:Nat, nil, '+, 100, 0) .
rewrites: 3
result ResultTriple: {'_+_['_*_['a.Nat,'b.Nat],'_*_['a.Nat,'c.Nat]],'Nat,
	'X:Nat <- '_+_['_*_['a.Nat,'b.Nat],'_*_['a.Nat,'c.Nat]]}
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_*_['a.Nat,'_+_['b.Nat,
    'c.Nat]], 'X:Nat, nil, '+, 100, 1) .
rewrites: 5
result ResultTriple?: (failure).ResultTriple?
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_+_['_*_['a.Nat,'b.Nat],
    '_*_['a.Nat,'c.Nat]], '_*_['X:Nat,'Y:Nat], nil, '*, 100, 0) .
rewrites: 3
result ResultTriple: {'_*_['a.Nat,'_+_['b.Nat,'c.Nat]],'Nat,
	'X:Nat <- 'a.Nat ; 
	'Y:Nat <- '_+_['b.Nat,'c.Nat]}
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_+_['_*_['a.Nat,'b.Nat],
    '_*_['a.Nat,'c.Nat]], '_*_['X:Nat,'Y:Nat], nil, '*, 100, 1) .
rewrites: 2
result ResultTriple: {'_*_['a.Nat,'_+_['b.Nat,'c.Nat]],'Nat,
	'X:Nat <- '_+_['b.Nat,'c.Nat] ; 
	'Y:Nat <- 'a.Nat}
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_+_['_*_['a.Nat,'b.Nat],
    '_*_['a.Nat,'c.Nat]], '_*_['X:Nat,'Y:Nat], nil, '*, 0, 0) .
rewrites: 2
result ResultTriple?: (failure).ResultTriple?
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_+_['_*_['a.Nat,'b.Nat],
    '_*_['a.Nat,'c.Nat],'_*_['a.Nat,'d.Nat]], '_*_['X:Nat,'Y:Nat], nil, '*,
    100, 0) .
rewrites: 9
result ResultTriple: {'_*_['a.Nat,'_+_['b.Nat,'c.Nat,'d.Nat]],'Nat,
	'X:Nat <- 'a.Nat ; 
	'Y:Nat <- '_+_['b.Nat,'c.Nat,'d.Nat]}
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_+_['_*_['a.Nat,'b.Nat],
    '_*_['a.Nat,'c.Nat],'_*_['a.Nat,'d.Nat]], '_*_['X:Nat,'Y:Nat], nil, '*,
    100, 1) .
rewrites: 2
result ResultTriple: {'_*_['a.Nat,'_+_['b.Nat,'c.Nat,'d.Nat]],'Nat,
	'X:Nat <- '_+_['b.Nat,'c.Nat,'d.Nat] ; 
	'Y:Nat <- 'a.Nat}
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_+_['_*_['a.Nat,'b.Nat],
    '_*_['a.Nat,'c.Nat],'_*_['a.Nat,'d.Nat]], '_*_['X:Nat,'Y:Nat], nil, '*, 1,
    0) .
rewrites: 8
result ResultTriple?: (failure).ResultTriple?
==========================================
reduce in META-LEVEL : metaSearch(['SEARCH-TEST], '_+_['_*_['a.Nat,'b.Nat],
    '_*_['a.Nat,'c.Nat],'_*_['a.Nat,'d.Nat]], '_*_['X:Nat,'Y:Nat], nil, '*, 1,
    1) .
rewrites: 8
result ResultTriple?: (failure).ResultTriple?
Maude> Bye.
